internal/backend/workloadproxy/reconciler.go

```diff
@@ -118,58 +103,47 @@ func (registry *Reconciler) Reconcile(cluster resource.ID, aliasToUpstreamAddres
 func (registry *Reconciler) ensureLB(cluster resource.ID, alias string, upstreamAddresses []string) error {
 	registry.logger.Log(registry.logLevel, "ensure LB", zap.String("cluster", cluster), zap.String("alias", alias), zap.Strings("upstream_addresses", upstreamAddresses))
 
-	hostPort := registry.hostPortForAlias(cluster, alias)
-	lbSts := registry.clusterToAliasToLBStatus[cluster][alias]
-
-	if lbSts == nil { // no LB yet, create and start it
-		tcpLB := &loadbalancer.TCP{
-			Logger: registry.lbLogger,
-			Proxy: tcpproxy.Proxy{
-				ListenFunc: registry.connProvider.Listen,
-			},
-			DialTimeout:    1 * time.Second,
-			TCPUserTimeout: 5 * time.Second,
-		}
+	if aliasToB := registry.clusterToAliasToBackend[cluster]; aliasToB == nil {
+		registry.clusterToAliasToBackend[cluster] = map[string]*upstream.List[*backend]{}
+	}
+
+	u := registry.clusterToAliasToBackend[cluster][alias]
+
+	if u == nil {
+		var err error
 
-		if err := tcpLB.AddRoute(
-			hostPort, upstreamAddresses,
-			upstream.WithHealthcheckTimeout(time.Second),
-			upstream.WithHealthcheckInterval(time.Minute),
+		if u, err = upstream.NewListWithCmp(
+			xiter.Map(func(addr string) *backend { return &backend{Addr: addr} }, slices.Values(upstreamAddresses)),
+			func(b *backend, b2 *backend) bool { return b.Addr == b2.Addr },
 		); err != nil {
-			return fmt.Errorf("failed to add route for %q/%q: %w", cluster, alias, err)
+			return err
 		}
+	} else {
+		u.Reconcile(xiter.Map(func(addr string) *backend { return &backend{Addr: addr} }, slices.Values(upstreamAddresses)))
+	}
 
-		if err := tcpLB.Start(); err != nil {
-			registry.logger.Log(registry.logLevel, "failed to start LB, attempt to stop it")
+	registry.clusterToAliasToBackend[cluster][alias] = u
 
-			startErr := fmt.Errorf("failed to start LB for %q/%q: %w", cluster, alias, err)
+	registry.aliasToCluster[alias] = cluster
 
-			// we still need to close the loadbalancer, so that the health checks goroutines get terminated
-			if closeErr := tcpLB.Close(); closeErr != nil {
-				return errors.Join(startErr, fmt.Errorf("failed to close LB: %w", closeErr))
-			}
+	return nil
+}
 
-			return startErr
-		}
+func (registry *Reconciler) removeLB(cluster resource.ID, alias string) {
```

Comment:
we should call this on some "idle" timeout - that is if we don't have DialContext for the cluster for some timeout, we can shut it down